pkg/specification/macspec/specification.go

```diff
@@ -0,0 +1,242 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package macspec provides access to LoRaWAN MAC specification-specific settings.
+package macspec
+
+import (
+	"fmt"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// macVersion returns the MACVersion as an integer.
+func macVersion(v ttnpb.MACVersion) int {
+	switch v {
+	case ttnpb.MACVersion_MAC_V1_0:
+		return 100
+	case ttnpb.MACVersion_MAC_V1_0_1:
+		return 101
+	case ttnpb.MACVersion_MAC_V1_0_2:
+		return 102
+	case ttnpb.MACVersion_MAC_V1_0_3:
+		return 103
+	case ttnpb.MACVersion_MAC_V1_0_4:
+		return 104
+	case ttnpb.MACVersion_MAC_V1_1:
+		return 110
+	default:
+		panic(fmt.Errorf("missed %q in macVersion", v))
+	}
+}
+
+// compareMACVersion compares MACVersions lhs to rhs:
+// -1 == lhs is less than rhs
+// 0 == lhs is equal to rhs
+// 1 == lhs is greater than rhs
+func compareMACVersion(lhs, rhs ttnpb.MACVersion) int {
+	switch r := macVersion(lhs) - macVersion(rhs); {
+	case r < 0:
+		return -1
+	case r == 0:
+		return 0
+	case r > 0:
+		return 1
+	default:
+		panic("unreachable")
+	}
+}
+
+// EncryptFOpts reports whether v requires MAC commands in FOpts to be encrypted.
+func EncryptFOpts(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
+
+// HasMaxFCntGap reports whether v defines a MaxFCntGap.
+func HasMaxFCntGap(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) < 0
+}
+
+// HasNoChangeTXPowerIndex reports whether v defines a no-change TxPowerIndex value.
+func HasNoChangeTXPowerIndex(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0
+}
+
+// HasNoChangeDataRateIndex reports whether v defines a no-change DataRateIndex value.
+func HasNoChangeDataRateIndex(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0
+}
+
+// IgnoreUplinksExceedingLengthLimit reports whether v requires Network Server to
+// silently drop uplinks exceeding selected data rate payload length limits.
+func IgnoreUplinksExceedingLengthLimit(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0 && compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// IncrementDevNonce reports whether v defines DevNonce as an incrementing counter.
+func IncrementDevNonce(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0
+}
+
+// UseNwkKey reports whether v uses a root NwkKey.
+func UseNwkKey(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
+
+// UseLegacyMIC reports whether v uses legacy MIC computation algorithm.
+func UseLegacyMIC(v ttnpb.MACVersion) bool {
```

Comment:
Legacy is a bit of an unfortunate name in The Things Stack. Initially, in 2018 when we started with The Things Stack, LoRaWAN 1.1 was the future and so LoRaWAN 1.0.x was legacy. That didn't really turn out to be the case.

With LoRaWAN 1.2, the MIC is going to change again. I think therefore that these shouldn't be boolean switches but rather per-version functions that compute the MIC. If we do that in `pkg/crypto`, do we need this function here?

This is not really a need for change now, but some foreseeable technical debt with LoRaWAN 1.2 on the horizon.

Reply:
We can have the equivalent of `crypto.EncryptionOption` for MIC - `crypto.MICOption` that basically controls how the MIC is generated, and then the `macspec` package can choose based on the LoRaWAN version what options to provide.

What is a bit different is that the LoRaWAN 1.1 MIC is basically derived from the 1.0.x MIC, so for now we're always computing the 1.0.x MIC. I guess we will see when 1.2 is out what kind of changes to introduce.

Reply:
The 1.1 MIC is already different, right, taking two integrity keys, ConfFCnt and TX parameters into account?

What will happen with 1.2 is that there comes a companion spec that specifies cipher suites. They will declare the MIC input and keys to use. Indeed let's see how that comes out of the washing machine and dryer and not address anything here and now.

Reply:
It is different but it acts as an extension of the 1.0.x MIC. So each time when you compute the 1.1 MIC, your first step is to actually compute the 'legacy' mic (which is 2 bytes), then you compute the 1.1 extension from there (the other two bytes).

This is an optimization used by the NS right now:
- You always compute the expected legacy MIC
- You always compare the expected legacy MIC with the provided legacy MIC. If they mismatch, you can already drop the frame since extending it with the 1.1 procedure won't make it match.
- If the end device is a 1.1 device, you actually compute the other two bytes and compare the MIC.